File: WhatsAppAutoReply/app/src/main/java/com/whatsappauto/reply/services/WhatsAppAccessibilityService.java
package com.whatsappauto.reply.services;

import android.accessibilityservice.AccessibilityService;
import android.app.Notification;
import android.app.PendingIntent;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityNodeInfo;

import androidx.core.view.accessibility.AccessibilityNodeInfoCompat;

import com.whatsappauto.reply.database.AppDatabase;
import com.whatsappauto.reply.database.AutoReplyDao;
import com.whatsappauto.reply.models.AutoReply;
import com.whatsappauto.reply.utils.WhatsAppHelper;

import java.util.List;
import java.util.concurrent.Executors;

public class WhatsAppAccessibilityService extends AccessibilityService {
    private static final String TAG = "WhatsAppAutoReply";
    private static final String WHATSAPP_PACKAGE = "com.whatsapp";
    
    private AutoReplyDao autoReplyDao;
    private String lastMessageSender = "";
    private String lastMessageText = "";

    @Override
    public void onCreate() {
        super.onCreate();
        AppDatabase db = AppDatabase.getInstance(this);
        autoReplyDao = db.autoReplyDao();
    }

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {
        if (event.getPackageName() == null || !event.getPackageName().toString().equals(WHATSAPP_PACKAGE)) {
            return;
        }

        switch (event.getEventType()) {
            case AccessibilityEvent.TYPE_NOTIFICATION_STATE_CHANGED:
                handleNotification(event);
                break;
            case AccessibilityEvent.TYPE_WINDOW_CONTENT_CHANGED:
                handleWindowContentChanged();
                break;
        }
    }

    private void handleNotification(AccessibilityEvent event) {
        Notification notification = (Notification) event.getParcelableData();
        if (notification != null) {
            Bundle extras = notification.extras;
            String title = extras.getString(Notification.EXTRA_TITLE);
            String text = extras.getString(Notification.EXTRA_TEXT);
            
            if (title != null && text != null) {
                checkAndReply(title, text, notification);
            }
        }
    }

    private void handleWindowContentChanged() {
        AccessibilityNodeInfo rootNode = getRootInActiveWindow();
        if (rootNode == null) return;

        // Extract messages from the chat window
        List<AccessibilityNodeInfo> messageNodes = rootNode.findAccessibilityNodeInfosByViewId("com.whatsapp:id/conversation_text_row");
        
        for (AccessibilityNodeInfo messageNode : messageNodes) {
            if (messageNode.getText() != null) {
                String messageText = messageNode.getText().toString();
                // Process the message
                processIncomingMessage(messageText);
            }
        }
    }

    private void processIncomingMessage(String messageText) {
        if (!messageText.equals(lastMessageText)) {
            lastMessageText = messageText;
            
            Executors.newSingleThreadExecutor().execute(() -> {
                List<AutoReply> autoReplies = autoReplyDao.findByKeyword(messageText);
                if (!autoReplies.isEmpty()) {
                    AutoReply autoReply = autoReplies.get(0);
                    sendReply(autoReply.getReplyMessage());
                }
            });
        }
    }

    private void checkAndReply(String sender, String message, Notification notification) {
        Executors.newSingleThreadExecutor().execute(() -> {
            List<AutoReply> autoReplies = autoReplyDao.findByKeyword(message);
            if (!autoReplies.isEmpty()) {
                AutoReply autoReply = autoReplies.get(0);
                
                // Open the notification
                try {
                    PendingIntent pendingIntent = notification.contentIntent;
                    if (pendingIntent != null) {
                        pendingIntent.send();
                        Thread.sleep(1000); // Wait for WhatsApp to open
                        
                        // Send the reply
                        sendReply(autoReply.getReplyMessage());
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error sending reply", e);
                }
            }
        });
    }

    private void sendReply(String replyMessage) {
        AccessibilityNodeInfo rootNode = getRootInActiveWindow();
        if (rootNode == null) return;

        // Find the message input field
        List<AccessibilityNodeInfo> messageFields = rootNode.findAccessibilityNodeInfosByViewId("com.whatsapp:id/entry");
        
        if (!messageFields.isEmpty()) {
            AccessibilityNodeInfo messageField = messageFields.get(0);
            
            // Set the text
            Bundle arguments = new Bundle();
            arguments.putCharSequence(AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, replyMessage);
            messageField.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, arguments);
            
            // Find and click the send button
            List<AccessibilityNodeInfo> sendButtons = rootNode.findAccessibilityNodeInfosByViewId("com.whatsapp:id/send");
            if (!sendButtons.isEmpty()) {
                sendButtons.get(0).performAction(AccessibilityNodeInfo.ACTION_CLICK);
            }
        }
    }

    public void sendScheduledMessage(String contactName, String message) {
        // This method will be called by the scheduler service
        WhatsAppHelper.openWhatsAppChat(this, contactName);
        
        // Wait for WhatsApp to open
        try {
            Thread.sleep(2000);
            sendReply(message);
        } catch (InterruptedException e) {
            Log.e(TAG, "Error in scheduled message", e);
        }
    }

    @Override
    public void onInterrupt() {
        Log.d(TAG, "Accessibility Service Interrupted");
    }
}

File: app/src/main/java/com/whatsappautoresponder/service/WhatsAppAccessibilityService.kt
package com.whatsappautoresponder.service

import android.accessibilityservice.AccessibilityService
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.accessibility.AccessibilityEvent
import android.view.accessibility.AccessibilityNodeInfo
import androidx.lifecycle.Observer
import com.whatsappautoresponder.database.AppDatabase
import com.whatsappautoresponder.database.AutoReplyRule
import com.whatsappautoresponder.utils.SharedPreferencesManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class WhatsAppAccessibilityService : AccessibilityService() {

    companion object {
        private const val TAG = "WhatsAppAccessibility"
        private const val WHATSAPP_PACKAGE = "com.whatsapp"
        var instance: WhatsAppAccessibilityService? = null
    }

    private lateinit var database: AppDatabase
    private lateinit var prefsManager: SharedPreferencesManager
    private val handler = Handler(Looper.getMainLooper())
    private val serviceScope = CoroutineScope(Dispatchers.IO)
    private var autoReplyRules: List<AutoReplyRule> = emptyList()

    override fun onServiceConnected() {
        super.onServiceConnected()
        instance = this
        database = AppDatabase.getInstance(this)
        prefsManager = SharedPreferencesManager(this)
        
        // Observe auto reply rules changes
        database.autoReplyRuleDao().getAllRules().observeForever { rules ->
            autoReplyRules = rules ?: emptyList()
        }
        
        Log.d(TAG, "WhatsApp Accessibility Service connected")
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent?) {
        if (!prefsManager.isAutoReplyEnabled()) return
        
        event?.let { accessibilityEvent ->
            when (accessibilityEvent.eventType) {
                AccessibilityEvent.TYPE_NOTIFICATION_STATE_CHANGED -> {
                    handleNotification(accessibilityEvent)
                }
                AccessibilityEvent.TYPE_WINDOW_CONTENT_CHANGED -> {
                    if (accessibilityEvent.packageName == WHATSAPP_PACKAGE) {
                        handleWindowContentChanged(accessibilityEvent)
                    }
                }
            }
        }
    }

    private fun handleNotification(event: AccessibilityEvent) {
        val packageName = event.packageName?.toString()
        if (packageName == WHATSAPP_PACKAGE) {
            val notification = event.text?.joinToString(" ")
            Log.d(TAG, "WhatsApp notification: $notification")
            
            notification?.let { notificationText ->
                processMessage(notificationText, null)
            }
        }
    }

    private fun handleWindowContentChanged(event: AccessibilityEvent) {
        if (event.source == null) return
        
        try {
            val rootNode = rootInActiveWindow ?: return
            val messageNodes = findMessageNodes(rootNode)
            
            messageNodes.forEach { messageNode ->
                val messageText = messageNode.text?.toString()
                val senderNode = findSenderNode(messageNode)
                val senderName = senderNode?.text?.toString()
                
                if (!messageText.isNullOrEmpty()) {
                    Log.d(TAG, "Message from $senderName: $messageText")
                    processMessage(messageText, senderName)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error processing window content", e)
        }
    }

    private fun findMessageNodes(rootNode: AccessibilityNodeInfo): List<AccessibilityNodeInfo> {
        val messageNodes = mutableListOf<AccessibilityNodeInfo>()
        
        // Look for message bubbles - these typically have specific class names or content descriptions
        fun traverse(node: AccessibilityNodeInfo) {
            // Check if this node contains a message
            if (isMessageNode(node)) {
                messageNodes.add(node)
            }
            
            for (i in 0 until node.childCount) {
                node.getChild(i)?.let { traverse(it) }
            }
        }
        
        traverse(rootNode)
        return messageNodes
    }

    private fun isMessageNode(node: AccessibilityNodeInfo): Boolean {
        val className = node.className?.toString()
        val viewId = node.viewIdResourceName
        
        // WhatsApp message bubbles typically have these characteristics
        return (className?.contains("TextView") == true && 
               (viewId?.contains("message") == true || 
                viewId?.contains("chat_message") == true ||
                node.contentDescription?.contains("message", ignoreCase = true) == true))
    }

    private fun findSenderNode(messageNode: AccessibilityNodeInfo): AccessibilityNodeInfo? {
        var parent = messageNode.parent
        while (parent != null) {
            // Look for sender name in parent or sibling nodes
            for (i in 0 until parent.childCount) {
                val child = parent.getChild(i)
                if (child != null && child != messageNode) {
                    val text = child.text?.toString()
                    if (!text.isNullOrEmpty() && text.length < 50 && !text.contains(":")) {
                        return child
                    }
                }
            }
            parent = parent.parent
        }
        return null
    }

    private fun processMessage(messageText: String, senderName: String?) {
        serviceScope.launch {
            try {
                // Find matching auto reply rule
                val matchingRule = autoReplyRules.find { rule ->
                    when {
                        rule.contactName.isNotEmpty() && senderName != null -> {
                            senderName.contains(rule.contactName, ignoreCase = true) &&
                            (rule.keyword.isEmpty() || messageText.contains(rule.keyword, ignoreCase = true))
                        }
                        rule.keyword.isNotEmpty() -> {
                            messageText.contains(rule.keyword, ignoreCase = true)
                        }
                        else -> false
                    }
                }

                matchingRule?.let { rule ->
                    Log.d(TAG, "Found matching rule: ${rule.keyword} -> ${rule.response}")
                    
                    // Apply delay if configured
                    val delay = prefsManager.getReplyDelay() * 1000L
                    if (delay > 0) {
                        Thread.sleep(delay)
                    }
                    
                    // Send auto reply
                    handler.post {
                        sendAutoReply(rule.response)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error processing message", e)
            }
        }
    }

    private fun sendAutoReply(response: String) {
        try {
            val rootNode = rootInActiveWindow ?: return
            val inputField = findInputField(rootNode)
            
            inputField?.let { field ->
                // Set text in input field
                val arguments = android.os.Bundle()
                arguments.putCharSequence(AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, response)
                field.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, arguments)
                
                // Find and click send button
                val sendButton = findSendButton(rootNode)
                sendButton?.performAction(AccessibilityNodeInfo.ACTION_CLICK)
                
                Log.d(TAG, "Auto reply sent: $response")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error sending auto reply", e)
        }
    }

    private fun findInputField(rootNode: AccessibilityNodeInfo): AccessibilityNodeInfo? {
        fun traverse(node: AccessibilityNodeInfo): AccessibilityNodeInfo? {
            val className = node.className?.toString()
            val viewId = node.viewIdResourceName
            
            // Look for WhatsApp message input field
            if (className == "android.widget.EditText" && 
                (viewId?.contains("entry") == true || 
                 viewId?.contains("input") == true ||
                 node.hint?.contains("message", ignoreCase = true) == true)) {
                return node
            }
            
            for (i in 0 until node.childCount) {
                node.getChild(i)?.let { child ->
                    traverse(child)?.let { return it }
                }
            }
            return null
        }
        
        return traverse(rootNode)
    }

    private fun findSendButton(rootNode: AccessibilityNodeInfo): AccessibilityNodeInfo? {
        fun traverse(node: AccessibilityNodeInfo): AccessibilityNodeInfo? {
            val className = node.className?.toString()
            val viewId = node.viewIdResourceName
            val contentDesc = node.contentDescription?.toString()
            
            // Look for WhatsApp send button
            if ((className?.contains("ImageButton") == true || className?.contains("Button") == true) &&
                (viewId?.contains("send") == true || 
                 contentDesc?.contains("send", ignoreCase = true) == true)) {
                return node
            }
            
            for (i in 0 until node.childCount) {
                node.getChild(i)?.let { child ->
                    traverse(child)?.let { return it }
                }
            }
            return null
        }
        
        return traverse(rootNode)
    }

    override fun onInterrupt() {
        Log.d(TAG, "WhatsApp Accessibility Service interrupted")
    }

    override fun onDestroy() {
        super.onDestroy()
        instance = null
        Log.d(TAG, "WhatsApp Accessibility Service destroyed")
    }

    // Public method to send scheduled messages
    fun sendScheduledMessage(contactName: String, message: String) {
        serviceScope.launch {
            try {
                // Open WhatsApp chat with specific contact
                openWhatsAppChat(contactName)
                
                // Wait for chat to load
                Thread.sleep(2000)
                
                // Send message
                handler.post {
                    sendAutoReply(message)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error sending scheduled message", e)
            }
        }
    }

    private fun openWhatsAppChat(contactName: String) {
        try {
            val intent = Intent(Intent.ACTION_VIEW).apply {
                setPackage(WHATSAPP_PACKAGE)
                putExtra("jid", "$contactName@s.whatsapp.net")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            startActivity(intent)
        } catch (e: Exception) {
            Log.e(TAG, "Error opening WhatsApp chat", e)
        }
    }
}